interpreter/cling/lib/MetaProcessor/InputValidator.cpp

```diff
@@ -15,191 +15,97 @@
+    if (m_ParenStack.empty()) {
+      MetaLexer::RAII RAII(Lex);
       Lex.Lex(Tok);
-      curPos = Lex.getLocation();
-      if (Tok.is(tok::ident)) {
-        if (Tok.getIdent()=="template")
-          m_ParenStack.push_back(tok::greater);
-      } else
-        curPos -= Tok.getLength(); // Rewind buffer for LexPunctuatorAndAdvance
+      if (Tok.is(tok::ident) && Tok.getIdent() == "template")
+        m_ParenStack.push_back(tok::less);
     }
 
     do {
-      lastKind = int(Tok.getKind());
-
-      const char* prevStart = curPos;
-      MetaLexer::LexPunctuatorAndAdvance(curPos, Tok);
-      const int kind = (int)Tok.getKind();
-
-      if (kind == commentTok) {
-        if (kind == tok::slash) {
-          if (multilineComment) {
-            // exiting a comment, unwind the stack
-            multilineComment = false;
-            commentTok = tok::slash;
-            unwindTokens(m_ParenStack, tok::slash);
-          }
-          // If we have a closing comment without a start it will be transformed
-          // to */; and clang reports an error for both the */ and the ;
-          // If we return kIncomplete, then just one error is printed, but too
-          // late: after the user has another expression which will always fail.
-          // So just deal with two errors for now
-          // else if (prevKind == tok::asterik) {
-          //  Res = kIncomplete;
-          // break;
-          // }
-          else   // wait for an asterik
-            commentTok = tok::asterik;
-        }
-        else {
-          assert(commentTok == tok::asterik && "Comment token not / or *");
-          if (!multilineComment) {
-            if ((curPos - prevStart) == 1) {
-              // entering a new comment
-              multilineComment = true;
-              m_ParenStack.push_back(tok::slash);
-            }
-          }
-          else // wait for closing / (must be next token)
-            commentTok = tok::slash;
-        }
-      }
-      else {
-        // If we're in a multiline, and waiting for the closing slash
-        // we gonna have to wait for another asterik first
-        if (multilineComment) {
-          if (kind == tok::eof) {
-            if (findBlockCommentEnd(prevStart, curPos))
-              unwindTokens(m_ParenStack, tok::slash);
-
-            // eof, were done anyway
+      if (Tok.isNot(tok::space))
+        lastNonSpaceTok = Tok;
+      Lex.Lex(Tok);
+
+      // In multiline comments, ignore anything that is not the `*/` token
+      if (inBlockComment() && Tok.isNot(tok::r_comment))
+        continue;
+
+      switch (Tok.getKind()) {
+      default:
+        break;
+      case tok::comment:
+        Lex.ReadToEndOfLine(Tok);
+        break;
+      case tok::r_comment:
+        if (inBlockComment())
+          m_ParenStack.pop_back();
+        else
+          Res = kMismatch;
+        break;
+
+      case tok::l_square: case tok::l_paren: case tok::l_brace:
```

Comment:
Do we want to add `LLVM_FALLTHROUGH` to silence potential warnings?

Reply:
IMHO, it should not be needed, as there are no statements interleaved with the case labels.
Thanks for the review, @vgvassilev!